src/webgpu/api/validation/createRenderPipeline.spec.ts

```diff
@@ -101,97 +101,122 @@ class F extends ValidationTest {
+  .fn(async t => {
+    const { isAsync } = t.params;
 
-  // Fail because lack of color states
-  const badDescriptor = t.getDescriptor({
-    targets: [],
-  });
+    const goodDescriptor = t.getDescriptor({
+      targets: [{ format: 'rgba8unorm' }],
+    });
+
+    // Control case
+    t.doCreateRenderPipelineTest(isAsync, true, goodDescriptor);
 
-  t.expectValidationError(() => {
-    t.device.createRenderPipeline(badDescriptor);
+    // Fail because lack of color states
+    const badDescriptor = t.getDescriptor({
+      targets: [],
+    });
+
+    t.doCreateRenderPipelineTest(isAsync, false, badDescriptor);
   });
-});
 
 g.test('color_formats_must_be_renderable')
-  .params(u => u.combine('format', kTextureFormats))
+  .params(u =>
+    u.combine('isAsync', [false, true]).combine('format', kTextureFormats as GPUTextureFormat[])
+  )
   .fn(async t => {
-    const format: GPUTextureFormat = t.params.format;
+    const { isAsync, format } = t.params;
     const info = kTextureFormatInfo[format];
     await t.selectDeviceOrSkipTestCase(info.feature);
 
     const descriptor = t.getDescriptor({ targets: [{ format }] });
 
     if (info.renderable && info.color) {
       // Succeeds when color format is renderable
-      t.device.createRenderPipeline(descriptor);
+      t.doCreateRenderPipelineTest(isAsync, true, descriptor);
     } else {
       // Fails because when format is non-renderable
-      t.expectValidationError(() => {
-        t.device.createRenderPipeline(descriptor);
-      });
+      t.doCreateRenderPipelineTest(isAsync, false, descriptor);
     }
   });
 
 g.test('sample_count_must_be_valid')
-  .paramsSimple([
-    { sampleCount: 0, _success: false },
-    { sampleCount: 1, _success: true },
-    { sampleCount: 2, _success: false },
-    { sampleCount: 3, _success: false },
-    { sampleCount: 4, _success: true },
-    { sampleCount: 8, _success: false },
-    { sampleCount: 16, _success: false },
-  ])
+  .params(u =>
+    u.combine('isAsync', [false, true]).combineWithParams([
+      { sampleCount: 0, _success: false },
+      { sampleCount: 1, _success: true },
+      { sampleCount: 2, _success: false },
+      { sampleCount: 3, _success: false },
+      { sampleCount: 4, _success: true },
+      { sampleCount: 8, _success: false },
+      { sampleCount: 16, _success: false },
+    ])
+  )
   .fn(async t => {
-    const { sampleCount, _success } = t.params;
+    const { isAsync, sampleCount, _success } = t.params;
 
     const descriptor = t.getDescriptor({ sampleCount });
 
-    if (_success) {
-      // Succeeds when sample count is valid
-      t.device.createRenderPipeline(descriptor);
-    } else {
-      // Fails when sample count is not 4 or 1
-      t.expectValidationError(() => {
-        t.device.createRenderPipeline(descriptor);
-      });
-    }
+    t.doCreateRenderPipelineTest(isAsync, _success, descriptor);
   });
 
 g.test('sample_count_must_be_equal_to_the_one_of_every_attachment_in_the_render_pass')
```

Comment:
While you are here, could you please leave a TODO, in the description (`.desc()`) of this test, to move it or deduplicate it with `attachment_compatibility.spec.ts`, because it is not actually testing validation in createRenderPipeline.

(Or if it is easy to fix this problem now, you can do that. I don't know how much work it will be.)

Reply:
Yes, I find it a little strange to expect validation error for commandEncoder.finish() here.

Reply:
I modified the render_pass_or_bundle_and_pipeline,sample_count test in attachment_compatibility.spec.ts to include the depthstencil texture cases, and remove this case from createRenderPipeline.spec.ts.